packages/server/src/fhir/repo.ts

```diff
@@ -2168,18 +2171,31 @@ export class Repository extends FhirRepository<PoolClient> implements Disposable
     callback: (client: PoolClient) => Promise<TResult>,
     options?: { serializable: boolean }
   ): Promise<TResult> {
-    try {
-      const client = await this.beginTransaction(options?.serializable ? 'SERIALIZABLE' : undefined);
-      const result = await callback(client);
-      await this.commitTransaction();
-      return result;
-    } catch (err) {
-      const operationOutcomeError = normalizeDatabaseError(err);
-      await this.rollbackTransaction(operationOutcomeError);
-      throw operationOutcomeError;
-    } finally {
-      this.endTransaction();
+    let error: OperationOutcomeError | undefined;
```

Comment:
ooo.  Clever.  This is a deeper insertion point than I was expecting...  But I like it, I think it makes a lot of sense.

I took a quick look over all of our uses of `withTransaction()` and asked the question "would we be ok if this retried 3 times?"

In all cases, I think it's good, and it will absolutely work.

The one case that I was somewhat uncertain about was the reindex worker, when processing a full page of 500 resources.  Even in that case, I think it's fine?  Might actually make the reindex job more robust?

Reply:
I struggled a bit going back and forth with handling this at the FHIR interaction level vs. here, but ultimately I think that having this handled consistently in one spot is the best outcome in terms of ensuring that the retry is performed correctly